src/test/java/org/elasticsearch/index/query/BaseQueryTestCase.java

```diff
@@ -0,0 +1,185 @@
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Ignore;
+import org.junit.Test;
+
+import java.io.IOException;
+
+@Ignore
+public abstract class BaseQueryTestCase<QB extends BaseQueryBuilder & Streamable> extends ElasticsearchTestCase {
+
+    private static Injector injector;
+    private static IndexQueryParserService queryParserService;
+    private static Index index;
+    protected QueryParseContext context;
+    protected XContentParser parser;
+
+    protected QB testQuery;
+
+    /**
+     * Setup for the whole base test class.
+     * @throws IOException
+     */
+    @BeforeClass
+    public static void init() throws IOException {
+        Settings settings = ImmutableSettings.settingsBuilder()
+                .put("name", BaseQueryTestCase.class.toString())
+                .put(IndexMetaData.SETTING_VERSION_CREATED, Version.CURRENT)
+                .build();
+
+        index = new Index("test");
+        injector = new ModulesBuilder().add(
+                new EnvironmentModule(new Environment(settings)),
+                new SettingsModule(settings),
+                new ThreadPoolModule(settings),
+                new IndicesQueriesModule(),
+                new ScriptModule(settings),
+                new IndexSettingsModule(index, settings),
+                new IndexCacheModule(settings),
+                new AnalysisModule(settings),
+                new SimilarityModule(settings),
+                new IndexNameModule(index),
+                new IndexQueryParserModule(settings),
+                new FunctionScoreModule(),
+                new AbstractModule() {
+                    @Override
+                    protected void configure() {
+                        bind(ClusterService.class).toProvider(Providers.of((ClusterService) null));
+                        bind(CircuitBreakerService.class).to(NoneCircuitBreakerService.class);
+                    }
+                }
+        ).createInjector();
+        queryParserService = injector.getInstance(IndexQueryParserService.class);
+    }
+
+    /**
+     * Setup for the individual tests, creating test query, context and XContentparser
+     */
+    @Before
+    @Override
+    public void setUp() throws Exception {
+        super.setUp();
+        context = new QueryParseContext(index, queryParserService);
+        testQuery = createRandomTestQuery();
+        String contentString = testQuery.toString();
+        parser = XContentFactory.xContent(contentString).createParser(contentString);
+        context.reset(parser);
+    }
+
+    /**
+     * Create a random query for the type that is being tested
+     * @return a randomized query
+     */
+    public abstract QB createRandomTestQuery();
+
+    /**
+     * Subclass should put assertions on the lucene query produced by the query builder under test here
+     * @param query
+     */
+    public abstract void doQueryAsserts(Query query) throws IOException;
```

Comment:
interesting questions... yea I think it's odd that we parse the query from string in `testToQuery`, I would just use `testQuery` straight-away, better test isolation. Also, instead of comparing `query` here with the `protected testQuery` coming from the parent class, I would make it more explicit what we have to compare by adding the `QueryBuilder` argument to the method and passing it in.